server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,296 @@
+                bulkShardRequest.shardId(),
+                bulkShardRequest.getRefreshPolicy(),
+                newItems
+            );
+            listener.onResponse(newBulkShardRequest);
+        };
+        try (var bulkItemReqRef = new RefCountingRunnable(onInferenceComplete)) {
+            BulkItemRequest[] items = bulkShardRequest.items();
+            for (int i = 0; i < items.length; i++) {
+                BulkItemRequest bulkItemRequest = items[i];
+                // Bulk item might be null because of previous errors, skip in that case
+                if (bulkItemRequest != null) {
+                    performInferenceOnBulkItemRequest(
+                        bulkItemRequest,
+                        fieldsForModels,
+                        i,
+                        onBulkItemFailure,
+                        failedItems,
+                        bulkItemReqRef.acquire()
+                    );
+                }
+            }
+        }
+    }
+
+    private void performInferenceOnBulkItemRequest(
+        BulkItemRequest bulkItemRequest,
+        Map<String, Set<String>> fieldsForModels,
+        Integer itemIndex,
+        BiConsumer<BulkItemRequest, Exception> onBulkItemFailure,
+        Set<Integer> failedItems,
+        Releasable releaseOnFinish
+    ) {
+
+        DocWriteRequest<?> docWriteRequest = bulkItemRequest.request();
+        Map<String, Object> sourceMap = null;
+        if (docWriteRequest instanceof IndexRequest indexRequest) {
+            sourceMap = indexRequest.sourceAsMap();
+        } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+            sourceMap = updateRequest.docAsUpsert() ? updateRequest.upsertRequest().sourceAsMap() : updateRequest.doc().sourceAsMap();
+        }
+        if (sourceMap == null || sourceMap.isEmpty()) {
+            releaseOnFinish.close();
+            return;
+        }
+        final Map<String, Object> docMap = new ConcurrentHashMap<>(sourceMap);
+
+        // When a document completes processing, update the source with the inference
+        try (var docRef = new RefCountingRunnable(() -> {
+            if (docWriteRequest instanceof IndexRequest indexRequest) {
+                indexRequest.source(docMap);
+            } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+                if (updateRequest.docAsUpsert()) {
+                    updateRequest.upsertRequest().source(docMap);
+                } else {
+                    updateRequest.doc().source(docMap);
+                }
+            }
+            releaseOnFinish.close();
+        })) {
+
+            for (Map.Entry<String, Set<String>> fieldModelsEntrySet : fieldsForModels.entrySet()) {
+                String modelId = fieldModelsEntrySet.getKey();
+
+                @SuppressWarnings("unchecked")
+                Map<String, Object> rootInferenceFieldMap = (Map<String, Object>) docMap.computeIfAbsent(
```

Comment:
How do we plan on handling cast errors here? It's theoretically possible for the user to provide a value for the `_semantic_text_inference` field with an invalid data type.

Reply:
I'd say we're safe as we're casting to Map<String, Object> - and the document source will be parsed already at this point. I don't think that this can fail if it's valid JSON (as should be at that stage) 🤔

Reply:
> and the document source will be parsed already at this point.

Parsed by whom?

There are three scenarios here:

 - User doing a "put/post" with the field already defined (obviously, hasn't been indexed already)
 - A Reindex occurring, obviously, this one is OK as it was indexed previously somewhere (hopefully by us :/)
 - A new document where this field exists or doesn't based on previously inferenced values (more than one field, meaning it doesn't exist for the first inference result but does for next).

This is the tricky part of having things that are generally mapper validates further up in the ingest pipeline.

So, we need to validate things are what we expect. We don't need to do a full parse of the internals (the mapper does this), but I am not sure blindly casting is wise.

Do we have tests covering the above scenarios I laid out yet?

Reply:
the variable `docMap` is retrieved using `sourceAsMap()` from the index or update request, which I believe invokes the XContentParser. So, it should be parsed at this point, right?

Reply:
I think the issue is that `_source` can be valid JSON without meeting the cast type expectations. For example:

```
{
  "_semantic_text_inference": "foo"
}
```

Reply:
You're right, for some reason I was not seeing that. Thanks for catching this!

Reply:
> from the index or update request, which I believe invokes the XContentParser. So, it should be parsed at this point, right?

It can be valid JSON (or CBOR, or SMILE), but it can 100% be invalid for what we care about.